# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


##### 1.
- $H_0$: $\mu$ = $\mu_0$ ⇒ $\mu$ = 123.7 (середній час до покупки не збільшився)
- $H_a$: $\mu$ > $\mu_0$ ⇒ $\mu$ > 123.7 (середній час до покупки збільшився)
- маємо правосторонній тест

In [1]:
import numpy as np
from scipy import stats

In [2]:
# Дані з умови

time = np.array([128, 135, 121, 142, 126, 151, 123])
m0 = 123.7
# n = 7
alpha = 0.10

In [3]:
# 2. Вибіркова статистика

x_bar = np.mean(time)
print(x_bar)
s = np.std(time, ddof=1)
print(s)
n = len(time)
print(n)

132.28571428571428
10.980502200328935
7


In [4]:
# 3. Критичне значення для правостороннього тесту

df = n - 1
t_crit = stats.t.ppf(1 - alpha, df)
t_crit

np.float64(1.4397557472577693)

In [5]:
#3. t-статистика

t_stat = (x_bar - m0) / (s / np.sqrt(n))
t_stat

np.float64(2.0687273144186475)

In [6]:
#3. p-value

p_value = 1 - stats.t.cdf(t_stat, df)
p_value

np.float64(0.042017654253682646)

In [7]:
#3. Або розрахунок t-статистики та p-value за допомогою бібліотеки scipy

res = stats.ttest_1samp(
    time,
    popmean=m0,
    alternative='greater'
)
res

TtestResult(statistic=np.float64(2.0687273144186475), pvalue=np.float64(0.042017654253682576), df=np.int64(6))

In [8]:
print("РЕЗУЛЬТАТИ ОБЧИСЛЕНЬ:")
print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"t-статистика: {t_stat:.3f}")
print(f"Критичне значення (df={df}): {t_crit:.3f}")
print(f"p-value: {p_value:.4f}")

РЕЗУЛЬТАТИ ОБЧИСЛЕНЬ:
Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
t-статистика: 2.069
Критичне значення (df=6): 1.440
p-value: 0.0420


In [9]:
print("Висновок:")
if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час до покупки користувачів більший за {m0} хв на рівні значущості {alpha}.")
else:
    print("Не відхиляємо H0: доказів недостатньо.")

Висновок:
Відхиляємо H0: середній час до покупки користувачів більший за 123.7 хв на рівні значущості 0.1.


In [10]:
# 4. Висновок на основі p-value

print("Висновок:")
if p_value < alpha:
    print(f"Відхиляємо H0: середній час до покупки користувачів більший за {m0} хв на рівні значущості {alpha}.")
else:
    print("Не відхиляємо H0: доказів недостатньо.")

Висновок:
Відхиляємо H0: середній час до покупки користувачів більший за 123.7 хв на рівні значущості 0.1.


**5. Висновок:** Редизайн сайту ймовірно продовжив час до покупки на рівні значущості 10 %.

In [11]:
# 6. Змінюємо рівень значущості на 0.05,

alpha = 0.05

# тоді перевизначаємо критичне значення

t_crit = stats.t.ppf(1 - alpha, df)
t_crit

np.float64(1.9431802805153022)

In [12]:
print("Висновок:")
if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час до покупки користувачів більший за {m0} хв на рівні значущості {alpha}.")
else:
    print("Не відхиляємо H0: доказів недостатньо.")

Висновок:
Відхиляємо H0: середній час до покупки користувачів більший за 123.7 хв на рівні значущості 0.05.


In [13]:
print("Висновок:")
if p_value < alpha:
    print(f"Відхиляємо H0: середній час до покупки користувачів більший за {m0} хв на рівні значущості {alpha}.")
else:
    print("Не відхиляємо H0: доказів недостатньо.")

Висновок:
Відхиляємо H0: середній час до покупки користувачів більший за 123.7 хв на рівні значущості 0.05.


**Висновок:** При зміні рівня значущості $\alpha = 0.05$ рішення щодо відхилення $H_0$ не змінюється.

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [14]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

In [15]:
# Дані з умови

p0 = 0.23
nobs = 1200
count = 311
alpha=0.03

- $H_0$: p = $p_0$ ⇒ p = 0.23 (частка дорослих, які впізнають логотип компанії не збільшилася)
- $H_a$: p > $p_0$ ⇒ p > 0.23 (частка дорослих, які впізнають логотип компанії збільшилася)
- маємо правосторонній тест

In [16]:
stat, pval = proportions_ztest(
    count,
    nobs,
    p0,
    alternative='larger'
)
stat, pval

(np.float64(2.3058316305222895), np.float64(0.010560021039195299))

In [17]:
print("ІНТЕРПРЕТАЦІЯ:")
if pval < alpha:
    print(f"Відхиляємо H0: Є статистично значущі докази, що частка дорослих, які впізнають логотип компанії, більша за {p0*100:.0f} %.")
else:
    print("Не відхиляємо H0: доказів недостатньо.")

ІНТЕРПРЕТАЦІЯ:
Відхиляємо H0: Є статистично значущі докази, що частка дорослих, які впізнають логотип компанії, більша за 23 %.


In [18]:
confint = proportion_confint(
    count,
    nobs,
    alpha,
    method='normal'
)
confint

(0.23171700302179205, 0.28661633031154127)

In [19]:
print("РЕЗУЛЬТАТИ ОБЧИСЛЕННЯ:")
print(f"z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97 % довірчий інтервал: {confint}")

РЕЗУЛЬТАТИ ОБЧИСЛЕННЯ:
z-статистика = 2.306
p-value = 0.0106
97 % довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


**Висновок:** З ймовірністю 97 % справжня частка дорослих, які впізнають логотип компанії після спеціальної рекламної кампанії, лежить у межах від 23.17 % до 28.66 %. Довірчий інтервал фактично повністю лежить вище 23 %, і відповідно узгоджується з результатом тесту: дані свідчать, що впізнаваність логотипу, найімовірніше, зросла.